### PR TITLE
JBIDE-20386: Autocompletion uses very outdated data (+ some more fixes)

### PR DESCRIPTION
**Pull request by ddekany**
Updated the set of directives (<#something>) and built-ins (exp?something) used for auto-completion and related-tag-highlighting, to be up to date with FreeMarker 2.3.24. It also fixes some related glitches and cleans up related code at some places; see details below.

What it doesn't do:
- It couldn't use the directive name and built-in name lists provided by the FreeMarker API, as the plugin needs more information than the names.
- It doesn't add camel case naming convention support. Another issue will have to be opened for that.

Change details:

- Added missing directive name Keyword-s, and the directives themselves to the Directive enum.

- Removed hard-coded auto-completion directive name list from AbstractDirective, and hard-coded (and already incomplete) end-tags from PartitionType. Instead, the Directive enum was extended to store all required information for these.

- Added directive-specific logic (for related-tag-highlighting and context variables) needed for some of the new directives: #attempt/#recover , and #else inside #list, #items.

- Created the BuiltIn enum (similar to the Directive enum in purpose) which contains the database of all the built-ins in 2.3.24. Modified BuiltInFragment auto-completion to build the left-hand-operand-class-specific sub-Maps from this enum.

- Added built-in auto-complation sub-Map for org.w3c.dom.Node values.

- Fixed broken instanceOf utility method used by auto-completion. (It often returned false instead of true.)

- Fixed oversight because of which the context variable class selection dialog didn't shown interfaces, only classes.

- Renamed IfElseDirective to ElseDirective (esp. as now it can occur under #list too).

- Tidied up default tag icon handling a bit: Added AbstractDirective.DEFAULT_IMAGE and related constructors.

### CODE REVIEW

File: plugins/org.jboss.ide.eclipse.freemarker/src/org/jboss/ide/eclipse/freemarker/model/AttemptDirective.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.ide.eclipse.freemarker.model;
```
Comment (alexeykazakov) [inline plugins/org.jboss.ide.eclipse.freemarker/src/org/jboss/ide/eclipse/freemarker/model/AttemptDirective.java]: Please use Eclipse Public License here and for other new sources instead.
Here is an example of header with EPL licence: https://github.com/jbosstools/jbosstools-openshift/blob/master/plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/core/internal/CDKConstantUtility.java

**Comment by alexeykazakov** [comment]
Please add junit tests.

File: plugins/org.jboss.ide.eclipse.freemarker/src/org/jboss/ide/eclipse/freemarker/model/AttemptDirective.java (hunk 1 above)
Comment (ddekany) [inline plugins/org.jboss.ide.eclipse.freemarker/src/org/jboss/ide/eclipse/freemarker/model/AttemptDirective.java]: Done (see commit).

**Comment by ddekany** [comment]
There weren't JUnit tests that covered the now updated/fixed functionality, and I wouldn't like to fix that too (the poor JUnit coverage) as part of this issue.

**Comment by alexeykazakov** [comment]
testPR

**Comment by alexeykazakov** [comment]
I see a test failure:
"junit.framework.AssertionFailedError: expected:<58> but was:<86>
	at junit.framework.Assert.fail(Assert.java:57)
	at junit.framework.Assert.failNotEquals(Assert.java:329)
	at junit.framework.Assert.assertEquals(Assert.java:78)
	at junit.framework.Assert.assertEquals(Assert.java:234)
	at junit.framework.Assert.assertEquals(Assert.java:241)
	at junit.framework.TestCase.assertEquals(TestCase.java:409)
	at org.jboss.ide.eclipse.freemarker.editor.coloring.test.AbstractColoringTest.validateColoring(AbstractColoringTest.java:72)
	at org.jboss.ide.eclipse.freemarker.editor.coloring.test.ListColoringTest.testListColoring(ListColoringTest.java:87)
"

**Comment by alexeykazakov** [comment]
@mickaelistria any idea why "testPR" doesn't trigger a test job?

**Comment by ddekany** [comment]
Fixed the list coloring test. (It seems that `mvn clean test` doesn't run all the tests... had to run `FreemarkerAllTests` manually.)

**Comment by alexeykazakov** [comment]
I used clean verify.

**Pull request merged**